httpgrpc/client.go

```diff
@@ -135,6 +134,7 @@ func (ch *Channel) NewStream(ctx context.Context, desc *grpc.StreamDesc, methodN
 	r, w := io.Pipe()
 	req, err := http.NewRequest("POST", reqUrlStr, r)
 	if err != nil {
+		cancel()
```

Comment:
should you also `defer cancel()` at the top of the go routine below?

Reply:
No, that would cancel the stream. But we have to return the stream -- it must survive beyond the scope of this function. Go vet was actually smart enough not to complain because _after_ this return, we do pass `cancel` to functions, which is what arranges for it to be eventually called. (The `newClientStream` will call it when the stream reaches EOF; and the finalizer below is that it to make sure it gets called, even if the caller that creates the stream fails to exhaust it.)

Reply:
But this code doesn't even work.. that `runtime.SetFinalizer` will never be invoked while `cs.doHttpCall` is running, because it's obviously pinning the reference.  Since you have to wait for `cs.doHttpCall` to exit anyway, why not just...

```
diff --git a/httpgrpc/client.go b/httpgrpc/client.go
index fc89e0d..a367fc1 100644
--- a/httpgrpc/client.go
+++ b/httpgrpc/client.go
@@ -12,7 +12,6 @@ import (
 	"net/textproto"
 	"net/url"
 	"path"
-	"runtime"
 	"strconv"
 	"strings"
 	"sync"
@@ -140,11 +139,13 @@ func (ch *Channel) NewStream(ctx context.Context, desc *grpc.StreamDesc, methodN
 	req.Header = h
 
 	cs := newClientStream(ctx, cancel, w, desc.ServerStreams, copts, ch.BaseURL)
-	// ensure that context is cancelled, even if caller
-	// fails to fully consume or cancel the stream
-	runtime.SetFinalizer(cs, func(*clientStream) { cancel() })
 
-	go cs.doHttpCall(ch.Transport, req)
+	go func() {
+		// ensure that context is cancelled, even if caller
+		// fails to fully consume or cancel the stream
+		defer cancel()
+		cs.doHttpCall(ch.Transport, req)
+	}()
 
 	return cs, nil
 }
```

Reply:
If you really, really need to actually use a finalizer (I don't understand the larger context here), then you'd need to wrap `*clientStream` in a delegating shim that can map GC -> context cancel, return the shim to the caller and not keep a reference to it from `doHttpCall`

Reply:
I'll have to take another look. I thought I already had some sort of wrapper to which only the caller got a reference, but maybe I made a mistake or forgot to do something here I've done in other similar situations.

Reply:
So it doesn't do any good to cancel _after_ `doHttpCall` completes because that is in fact the very thing we are trying to cancel: if the RPC is aborted, we want to cancel the outstanding HTTP request instead of leaving that connection open and pending indefinitely.

The finalizer is the only correct way to arrange this AFAICT. But you are correct that I put the finalizer on the wrong value -- a value that is pinned to the heap by the very goroutine it wants to cancel. The fix is in #50.